File: backend/app/schemas/dictionary.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class DictionaryTypeBase(BaseModel):
    type_code: str = Field(min_length=1, max_length=100)
    type_name: str = Field(min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    sort_order: int = 0


class DictionaryTypeCreate(DictionaryTypeBase):
    pass


class DictionaryTypeUpdate(BaseModel):
    type_code: str | None = Field(default=None, min_length=1, max_length=100)
    type_name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    sort_order: int | None = None


class DictionaryTypeListFilter(BaseModel):
    type_code: str | None = None
    type_name: str | None = None


class DictionaryTypePublic(DictionaryTypeBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime | None

    model_config = {"from_attributes": True}


class DictionaryTypesPublic(BaseModel):
    data: list[DictionaryTypePublic]
    count: int


class DictionaryItemBase(BaseModel):
    item_code: str = Field(min_length=1, max_length=100)
    item_label: str = Field(min_length=1, max_length=200)
    item_value: str = Field(min_length=1, max_length=500)
    sort_order: int = 0
    enabled: bool = True


class DictionaryItemCreate(DictionaryItemBase):
    type_id: uuid.UUID


class DictionaryItemUpdate(BaseModel):
    type_id: uuid.UUID | None = None
    item_code: str | None = Field(default=None, min_length=1, max_length=100)
    item_label: str | None = Field(default=None, min_length=1, max_length=200)
    item_value: str | None = Field(default=None, min_length=1, max_length=500)
    sort_order: int | None = None
    enabled: bool | None = None


class DictionaryItemListFilter(BaseModel):
    type_id: uuid.UUID | None = None
    item_code: str | None = None
    item_label: str | None = None
    enabled: bool | None = None


class DictionaryItemPublic(DictionaryItemBase):
    id: uuid.UUID
    type_id: uuid.UUID
    created_at: datetime
    updated_at: datetime | None

    model_config = {"from_attributes": True}


class DictionaryItemsPublic(BaseModel):
    data: list[DictionaryItemPublic]
    count: int


class DictionaryOptionItemPublic(BaseModel):
    code: str
    label: str
    value: str
    sort_order: int


class DictionaryOptionsPublic(BaseModel):
    type_code: str
    data: list[DictionaryOptionItemPublic]

File: frontend/src/api/dictionary.ts
import api from '@/api'

export interface DictionaryTypePublic {
  id: string
  type_code: string
  type_name: string
  description: string | null
  sort_order: number
  created_at: string
  updated_at: string | null
}

export interface DictionaryTypesPublic {
  data: DictionaryTypePublic[]
  count: number
}

export interface DictionaryTypeListParams {
  type_code?: string
  type_name?: string
  skip?: number
  limit?: number
}

export interface DictionaryTypeCreate {
  type_code: string
  type_name: string
  description?: string | null
  sort_order?: number
}

export interface DictionaryTypeUpdate {
  type_code?: string
  type_name?: string
  description?: string | null
  sort_order?: number
}

export interface DictionaryItemPublic {
  id: string
  type_id: string
  item_code: string
  item_label: string
  item_value: string
  sort_order: number
  enabled: boolean
  created_at: string
  updated_at: string | null
}

export interface DictionaryItemsPublic {
  data: DictionaryItemPublic[]
  count: number
}

export interface DictionaryItemListParams {
  type_id?: string
  item_code?: string
  item_label?: string
  enabled?: boolean
  skip?: number
  limit?: number
}

export interface DictionaryItemCreate {
  type_id: string
  item_code: string
  item_label: string
  item_value: string
  sort_order?: number
  enabled?: boolean
}

export interface DictionaryItemUpdate {
  type_id?: string
  item_code?: string
  item_label?: string
  item_value?: string
  sort_order?: number
  enabled?: boolean
}

export interface DictionaryOptionItemPublic {
  code: string
  label: string
  value: string
  sort_order: number
}

export interface DictionaryOptionsPublic {
  type_code: string
  data: DictionaryOptionItemPublic[]
}

export const dictionaryApi = {
  listTypes: (params?: DictionaryTypeListParams) =>
    api.get<DictionaryTypesPublic>('/dictionaries/', { params }),

  getType: (id: string) => api.get<DictionaryTypePublic>(`/dictionaries/${id}`),

  createType: (data: DictionaryTypeCreate) =>
    api.post<DictionaryTypePublic>('/dictionaries/', data),

  updateType: (id: string, data: DictionaryTypeUpdate) =>
    api.patch<DictionaryTypePublic>(`/dictionaries/${id}`, data),

  deleteType: (id: string) => api.delete(`/dictionaries/${id}`),

  listItems: (params?: DictionaryItemListParams) =>
    api.get<DictionaryItemsPublic>('/dictionary-items/', { params }),

  getItem: (id: string) => api.get<DictionaryItemPublic>(`/dictionary-items/${id}`),

  createItem: (data: DictionaryItemCreate) =>
    api.post<DictionaryItemPublic>('/dictionary-items/', data),

  updateItem: (id: string, data: DictionaryItemUpdate) =>
    api.patch<DictionaryItemPublic>(`/dictionary-items/${id}`, data),

  deleteItem: (id: string) => api.delete(`/dictionary-items/${id}`),

  getEnabledItemsByTypeCode: (typeCode: string) =>
    api.get<DictionaryOptionsPublic>(`/dictionaries/by-code/${typeCode}/items`),
}
